database/test/test_services.py

```diff
@@ -157,3 +165,198 @@ def test_populate_local_koodistot(populate_local_koodistot, main_db_params):
                     assert code_count > 0
     finally:
         conn.close()
+
+
+@pytest.fixture()
+def validate_invalid_plan(ryhti_client_url, complete_test_plan):
+    """
+    Validate an invalid Ryhti plan against the Ryhti API. Complete test plan is not yet a
+    valid plan, because it contains test code values.
+
+    An invalid plan should make lambda return http 200 OK (to indicate that the validation
+    has been run successfully), with the validation errors returned in the payload.
+    """
+    payload = {"event_type": 1, "save_json": True}
+    r = requests.post(ryhti_client_url, data=json.dumps(payload))
+    data = r.json()
+    print(data)
+    assert data["statusCode"] == 200
+    assert data["title"] == "Plan validations run."
+    assert (
+        data["details"][complete_test_plan.id]
+        == f"Validation FAILED for {complete_test_plan.id}."
+    )
+    # our invalid plan has invalid code uri value. For some reason, the Ryhti API
+    # currently considers this as a JSON deserialization error (which it is not),
+    # so it returns HTTP 400 instead of 422.
+    assert data["ryhti_responses"][complete_test_plan.id]["status"] == 400
+    assert data["ryhti_responses"][complete_test_plan.id]["errors"]
+
+
+def test_validate_invalid_plan(validate_invalid_plan, main_db_params):
+    """
+    Test the whole lambda endpoint with an invalid plan
+    """
+    conn = psycopg2.connect(**main_db_params)
+    try:
+        with conn.cursor() as cur:
+            cur.execute(f"SELECT validated_at, validation_errors FROM hame.plan")
+            validation_date, errors = cur.fetchone()
+            assert validation_date
+            assert errors
+    finally:
+        conn.close()
+
+
+@pytest.fixture()
+def valid_plan_in_preparation(
+    session: Session,
+    populate_koodistot: None,
```

Comment:
@msorvoja I think populate_koodistot still needs to be run (though its output is None), because without it, there will be no codes in the database to use in the queries below! :)